core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -371,6 +371,14 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
+    // System property spark.yarn.app.id must be set if user code ran by AM on a YARN cluster
+    // yarn-standalone is deprecated, but still supported
+    if ((master == "yarn-cluster" || master == "yarn-standalone") &&
+        !_conf.contains("spark.yarn.app.id")) {
```

Comment:
It's necessary because that's what differentiates "user instantiating SparkContext in yarn-cluster mode" and "ApplicationMaster running user code that instantiates SparkContext in yarn-cluster mode". If you remove that check, yarn-cluster mode will always fail this check. See this code in ApplicationMaster:

        // Set the master property to match the requested mode.
        System.setProperty("spark.master", "yarn-cluster")